### PR TITLE
Add/navigation blocks post processing after migration from menu items

### PR DESCRIPTION
**Pull request by adamziel**
## Description
Related to https://github.com/WordPress/gutenberg/issues/36616

**Problem:** 
1. Polylang (and many other plugins) has a special menu item called "Language switcher". Whenever a menu is rendered or returned by an API endpoint, that menu item is turned into many items like "English, French, Spanish" by a filter..
1. Gutenberg migrates menu items to blocks one by one based on the REST API output. Language switcher is represented as many links, so instead of ending up with a single "language switcher" block, it ends up with one navigation link block per language.

**Proposed solution:**
Add a set od filters to postprocess the `menu items -> blocks` conversion results. A usage example for the Polylang plugin would look like this:

Sample usage in JS (runs after selecting a classic menu in the block):
```js
wp.hooks.addFilter(
	'navigation.menuItemsToBlocks',
	'polylang/include-language-picker',
	( navBlock, menuItems ) => ({
		...navBlock,
		innerBlocks: mapBlockTree(
			navBlock.innerBlocks,
			block => {
				if(block.name === "core/navigation-link" && block.attributes?.url === "#pll_switcher") {
					return createBlock( 'polylang/navigation-language-switcher' );
				}
				return block;
			}
		)
	})
);

function mapBlockTree(blocks, mapper) {
    return blocks.map(block => ({
        ...mapper(block),
        innerBlocks: mapBlockTree(block.innerBlocks, mapper)
    }));
}
```

A corresponding filter would have to be implemented in PHP to catch the conversion that runs when switching to a block theme.

As a side note, the migration could potentially be handled entirely on the PHP side instead of being implemented in both PHP and JS – but that's for another day. cc @noisysocks @talldan @getdave @manooweb 

## How has this been tested?

Confirm that all the checks are green.

### CODE REVIEW

**Comment by manooweb** [comment]
Thanks a lot @adamziel to have been work on it 👍 and the lead.
I tested on my side and it seems the filter added in js code is sufficient at this step of my test. 

However, I already found a mistake in the sample code provided

```js
	return {
		name: "navigation-language-switcher",
		clientId: block.clientId,
	};

```
We don't return a block here and it triggers an error later in the execution

```
serializer.js:216 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'attributes')
    at getCommentAttributes (serializer.js:216)
    at serializeBlock (serializer.js:357)
    at serializer.js:393
    at Array.map (<anonymous>)
    at serialize (serializer.js:393)
    at use-create-navigation-menu.js:27
    at onFinishMenuCreation (index.js:105)
    at index.js:127
    at index.js:158
    at invokePassiveEffectCreate (react-dom.development.js:23517)

```
I simply replaced it by a call to `@wordpress/blocks createBlock()` method like

```js
	if( block.name === "core/navigation-link" && block.attributes?.url === "#pll_switcher" ) {
		return createBlock( 'polylang/navigation-language-switcher' );
	}

```
for example

It works fine like that and now it's the language switcher block which appears as a child of the navigation block
image

However the language switcher block attributes isn't correctly set because the language switcher menu item options are a custom post meta which isn't returned by default.

I will take a look at the `navigation_after_parse_blocks_from_menu_items` added to `lib/navigation.php` if it allow us to solve this problem.

I will let you know

**Comment by adamziel** [comment]
@manooweb you're bringing up a good point here about the custom attributes. I think the menu migration would have to PHP-only to support that. I imagine that would involve a new endpoint that would run the migration code that's already there. By the way, @talldan mentioned removing the PHP filter that runs on theme switch in https://github.com/WordPress/gutenberg/pull/36604#issuecomment-983217397, which adds urgency to this issue.

> However, I already found a mistake in the sample code provided

Good catch, thanks! I just updated that snippet

**Comment by manooweb** [comment]
> @manooweb you're bringing up a good point here about the custom attributes. I think the menu migration would have to PHP-only to support that. I imagine that would involve a new endpoint that would run the migration code that's already there. By the way, @talldan mentioned removing the PHP filter that runs on theme switch in [#36604 (comment)](https://github.com/WordPress/gutenberg/pull/36604#issuecomment-983217397), which adds urgency to this issue.
> 
@adamziel it seems to me that it isn't the context we're using.
Indeed menu items are retrieved by a REST request with the new menu-items endpoint https://github.com/WordPress/gutenberg/blob/v12.1.0-rc.1/lib/class-wp-rest-menu-items-controller.php#L34-L39

I'm going to search in that direction instead. We're used to use the REST API for the language for example and it seems not to be difficult to push our custom post meta here.

File: lib/navigation.php
Comment (spacedmonkey) [inline lib/navigation.php]: This filter will need documentation

Comment (spacedmonkey) [reply lib/navigation.php]: The filter is a little long as well, can we think something a little shorter.

**Comment by manooweb** [comment]
On my side, I studied a little bit further espacially because we didn't get the Polylang language switcher menu item meta ([_pll_menu_item](https://github.com/polylang/polylang/blob/master/admin/admin-nav-menu.php#L150-L163)).

So I register the meta to be able to get it when the [menu-items](https://github.com/WordPress/gutenberg/blob/v12.1.0-rc.1/lib/class-wp-rest-menu-items-controller.php#L30-L49) endpoint is called during the conversion process.

```php
<?php
add_action( 'rest_api_init', 'pll_register_switcher_menu_item_options_meta_rest_field' );

function pll_register_switcher_menu_item_options_meta_rest_field() {
    $return = register_meta(
        'post',
        '_pll_menu_item',
        array(
            'object_subtype' => 'nav_menu_item',
            'description' => __( 'Language switcher menu item options.', 'polylang-pro' ),
            'single' => true,
            'show_in_rest' => array(
                'schema' => array(
                    'type' => 'object',
                    'additionalProperties' => array(
                        'type' => 'integer',
                    ),
                )
            ),
        )
    );
}

```
and I got correctly the meta with the menu-items REST API call.

```json
{
    "id": 585,
    "title": {
        "raw": "Langues",
        "rendered": "Langues"
    },
    "status": "publish",
    "url": "#pll_switcher",
    "attr_title": "",
    "description": "",
    "type": "custom",
    "type_label": "Liste des langues",
    "object": "custom",
    "object_id": 585,
    "content": {
        "raw": "",
        "rendered": "",
        "block_version": 0
    },
    "parent": 0,
    "menu_order": 1,
    "target": "",
    "classes": [
        ""
    ],
    "xfn": [
        ""
    ],
    "invalid": false,
    "meta": {
        "_pll_menu_item": {
            "hide_if_no_translation": 0,
            "hide_current": 0,
            "force_home": 0,
            "show_flags": 1,
            "show_names": 1,
            "dropdown": 0
        }
    },
    "menus": 66,
    "_links": {
        "self": [
            {
                "href": "http://support.local/wp-json/wp/v2/menu-items/585"
            }
        ],
        "collection": [
            {
                "href": "http://support.local/wp-json/wp/v2/menu-items"
            }
        ],
        "about": [
            {
                "href": "http://support.local/wp-json/wp/v2/types/nav_menu_item"
            }
        ],
        "wp:term": [
            {
                "taxonomy": "nav_menu",
                "embeddable": true,
                "href": "http://support.local/wp-json/wp/v2/menus?post=585"
            }
        ],
        "wp:action-publish": [
            {
                "href": "http://support.local/wp-json/wp/v2/menu-items/585"
            }
        ],
        "wp:action-unfiltered-html": [
            {
                "href": "http://support.local/wp-json/wp/v2/menu-items/585"
            }
        ],
        "wp:action-create-menus": [
            {
                "href": "http://support.local/wp-json/wp/v2/menu-items/585"
            }
        ],
        "wp:action-assign-menus": [
            {
                "href": "http://support.local/wp-json/wp/v2/menu-items/585"
            }
        ],
        "curies": [
            {
                "name": "wp",
                "href": "https://api.w.org/{rel}",
                "templated": true
            }
        ]
    }
}

```
Then I adapted the code provided, to pass the meta to the block creation
NB: `find` is the lodash function of course.

```js
function mapBlockTree( blocks, menuItems, mapper ) {
	return blocks.map(
		block => (
			{
				...mapper( block, menuItems ),
				innerBlocks: mapBlockTree( block.innerBlocks, menuItems, mapper )
			}
		)
	 );;
}

addFilter(
	'navigation.menuItemsToBlocks',
	'polylang/include-language-switcher',
	( blocks, menuItems ) => (
		{
			...blocks,
			innerBlocks: mapBlockTree(
				blocks.innerBlocks,
				menuItems,
				( block, menuItems ) => {
					if( block.name === "core/navigation-link" && block.attributes?.url === "#pll_switcher" ) {
						const menuItem = find( menuItems, { url: '#pll_switcher' } ); // Get the corresponding menu item.
						const attributes = menuItem.meta._pll_menu_item; // Get its options.
						return createBlock( navigationLanguageSwitcherName, attributes );
					}
					return block;
				}
			)
	 	}
	)
);
```
However, even if the menu item got the correct meta during the process
image

The block attributes UI didn't display the right options. `Display flags` should be checked because the `show_flags` meta option is set to 1

image

I noticed that the `wp_navigation` post seems good to me in DB
image

image

So I thought because the navigation block mapping wasn't updated. Then I adapted the code again to take the blocks mapping with the new language switcher block client id.

```js
function mapBlockTree( blocks, menuItems, blocksMapping, mapper ) {
	return blocks.map(
		block => (
			{
				...mapper( block, menuItems, blocksMapping ),
				innerBlocks: mapBlockTree( block.innerBlocks, menuItems, blocksMapping, mapper )
			}
		)
	 );;
}

addFilter(
	'navigation.menuItemsToBlocks',
	'polylang/include-language-switcher',
	( blocks, menuItems ) => (
		{
			...blocks,
			innerBlocks: mapBlockTree(
				blocks.innerBlocks,
				menuItems,
				blocks.mapping,
				( block, menuItems, blocksMapping ) => {
					if( block.name === "core/navigation-link" && block.attributes?.url === "#pll_switcher" ) {
						const menuItem = find( menuItems, { url: '#pll_switcher' } ); // Get the corresponding menu item.
						const attributes = menuItem.meta._pll_menu_item; // Get its options.
						const newBlock = createBlock( navigationLanguageSwitcherName, attributes );
						blocksMapping[ menuItem.id ] = newBlock.clientId;
						return newBlock;
					}
					return block;
				}
			)
	 	}
	)
);

```

Unfortunately it changed nothing to the result the language switcher block attributes aren't right in the UI.

I don't see how to figure out this issue and where to search.

If you have an idea I'm very interested in

Many thanks

**Comment by adamziel** [comment]
@manooweb somehow I can't find the `polylang/navigation-language-switcher` anywhere which means I can't test. Was it released with the latest PolyLang version?

Anyway, let me take my best guess:
Block attributes tend to use camelCase, e.g. `showFlags`. However, the API returns snake-cased data, e.g. `show_flags`. Based on your debugger screenshot, I believe things are still snake-cased when we reach this fragment:  

```js
const attributes = menuItem.meta._pll_menu_item; // Get its options.
const newBlock = createBlock( navigationLanguageSwitcherName, attributes );
```

And so the call unwinds to
```js
const newBlock = createBlock(
	"polylang/navigation-language-switcher",
	{
		dropdown: 0,
		force_home: 0,
		hide_current: 0,
		hide_if_no_translation: 0,
		show_flags: 1,
		show_names: 1,
	}
);
```

When I think you're after:

```js
const newBlock = createBlock(
	"polylang/navigation-language-switcher",
	{
		dropdown: 0,
		forceHome: 0,
		hideCurrent: 0,
		hideIfNoTranslation: 0,
		showFlags: 1,
		showNames: 1,
	}
);
```

LMK if that helped.

**Comment by manooweb** [comment]
> @manooweb somehow I can't find the `polylang/navigation-language-switcher` anywhere which means I can't test. Was it released with the latest PolyLang version?
> 

I'm aware it's difficult to understand without this part of the code but it is developed in the premium version only.
The language switcher block exists since several versions now.
I'm going to try to provide a piece of code to work with its free version.

> Anyway, let me take my best guess: Block attributes tend to use camelCase, e.g. `showFlags`. However, the API returns snake-cased data, e.g. `show_flags`. Based on your debugger screenshot, I believe things are still snake-cased when we reach this fragment:
> 
> ```js
> const attributes = menuItem.meta._pll_menu_item; // Get its options.
> const newBlock = createBlock( navigationLanguageSwitcherName, attributes );
> ```
> 
> And so the call unwinds to
> 
> ```js
> const newBlock = createBlock(
> 	"polylang/navigation-language-switcher",
> 	{
> 		dropdown: 0,
> 		force_home: 0,
> 		hide_current: 0,
> 		hide_if_no_translation: 0,
> 		show_flags: 1,
> 		show_names: 1,
> 	}
> );
> ```
> 
> When I think you're after:
> 
> ```js
> const newBlock = createBlock(
> 	"polylang/navigation-language-switcher",
> 	{
> 		dropdown: 0,
> 		forceHome: 0,
> 		hideCurrent: 0,
> 		hideIfNoTranslation: 0,
> 		showFlags: 1,
> 		showNames: 1,
> 	}
> );
> ```
> 
> LMK if that helped.

I don't think it's a case problem with the attribute names. The block itself works correctly with these attributes in snake case.
image

image

We develop like that to be consistent with the language switcher in PHP code espacially to maintain these options only in one place.
https://github.com/polylang/polylang/blob/3.1.2/include/switcher.php#L45-L52

**Comment by manooweb** [comment]
@adamziel I extracted in a GH repository the code necessary to have the language switcher block as a stand alone block or as a child of the `core/navigation` block.

You have a readme file to explain how to install it and how to use it with the second use case that we are interested in here.

https://github.com/manooweb/gutenberg-block-navigation-conversion-test

If you need any other explanation or have issues, LMK 😉

Thanks

**Comment by adamziel** [comment]
@manooweb  thanks! I identified the problem. The attributes are registered as `boolean`, but the data from the API is of type `number`. When I cast it to bool like `attributes.dropdown = !! attributes.dropdown`, it all works as expected. I suspect this is related to the block parser sanitizing the navigation post `content` it receives in response to the POST API request. Let me know if that helps.

**Comment by manooweb** [comment]
> @manooweb thanks! I identified the problem. The attributes are registered as `boolean`, but the data from the API is of type `number`. When I cast it to bool like `attributes.dropdown = !! attributes.dropdown`, it all works as expected. I suspect this is related to the block parser sanitizing the navigation post `content` it receives in response to the POST API request. Let me know if that helps.



@adamziel Yes!! That helps 👍 Thanks

**Comment by manooweb** [comment]
@adamziel LGTM
I adapted our code like this https://github.com/manooweb/gutenberg-block-navigation-conversion-test/blob/ca02378870c8f69f49fd79a6d40645eb1e536e01/language-switcher-blocks.js#L157-L158

And now the result is as expected

From the classic menu, I checked `Display flags` option
image

and the conversion to the corresponding block in the `core/navigation` block works correctly
image

**Edit:** in fact it depends on how to declare REST API schema to retrieve the meta. I had declared `integer` because it's the real type saved in DB however we need to get a boolean value that corresponds really on its use.

So I reverted the JS code and I corrected the `register_meta` call with the correct type in the REST API schema.
https://github.com/manooweb/gutenberg-block-navigation-conversion-test/blob/acd27ef464f3bb1013dde0df67e3022e5c1ecdfd/block-editor-switcher-block.php#L327

Do you think this PR can be merged in the next release and available in the next WordPress release?

Many thanks for the work and time to help 👍😉

**Comment by adamziel** [comment]
> Do you think this PR can be merged in the next release and available in the next WordPress release?

@manooweb I believe so! I just need an approval here :-) @getdave @noisysocks @talldan @draganescu @Mamaduka anyone willing to approve ?

**Review by getdave — APPROVED** [review]
Looks good and it's as I imagined this would be solved. Nice work!

I think we need to document this filter as I imagine folks would want to use it. Not sure where the best place for that is. Maybe @mkaz would know?

**Comment by draganescu** [comment]
I think this is a smart solution to the problem but I have a few, potentially convoluted, concerns:

Will we ever want to transform classic menus into something else other than submenu and link? Adding this filter will be a contract that we always return submenu and link to extenders.
- will a future deprecation of the edit component of the navigation block continue to work when plugins use this filter?

This is a 1st in the block library :)
- is this all we need to do (add a filter) considering no other core block needs any filter other than the standard block filters?

Looking at the problem:
> Language switcher is represented as many links, so instead of ending up with a single "language switcher" block, it ends up with one navigation link block per language.

- Is is possible to make the filter be less low level, as what seems to be needed is a way to insert a special block (the langauge switcher block) in the navigation edit component made of some of the links already found in it (the language links)? A sort of auto-transform thing?

All my concerns are summed up by just being worried that by adding a filter on client side conversion we make a contract about the structure of the navigation block to the world.

**Comment by getdave** [comment]
I will defer to @adamziel who no doubt is better placed to answer than I. Nonetheless:

> Will we ever want to transform classic menus into something else other than submenu and link? Adding this filter will be a contract that we always return submenu and link to extenders.

I don't think we are forcing users to return a particular structure. We're giving them the option to hook into each menn item and change how that is transformed into a block. If they want menu item X to turn into multiple blocks equivalents then that's their choice. They could just as easily do nothing or just subtly change the conversion. Indeed Adam's example shows you can just `return block` and do nothing.

> All my concerns are summed up by just being worried that by adding a filter on client side conversion we make a contract about the structure of the navigation block to the world.

We are making a contract on that filter. I felt it was safe to assume that we will always want to provide a means of transforming menu items to blocks. If we're not 100% sure that's not going to be a feature then perhaps we need to make this filter experimental for a while and make a note to promote to first class only after a few Gutenberg Plugin versions have elapsed.

**Comment by manooweb** [comment]
@draganescu, today we can extend "classic" menu with custom menu item and the current version of `core/navigation` block convert all the menu item from an existing "classic" menu only to a `core/navigation-link` which isn't the result expected.

We need to change the conversion behaviour to catch the custom menu item and change its conversion to another block than  the default `core/navigation-link` block and then have the same render as we're using the initial "classic" menu.

File: packages/block-library/src/navigation/menu-items-to-blocks.js
Comment (draganescu) [inline packages/block-library/src/navigation/menu-items-to-blocks.js]: @adamziel @getdave if the filter would allow extenders to provide a drop in replacement to the default converter `mapMenuItemsToBlocks`, one starting from `menuTree`, instead of a way to change the already converted block tree, wouldn't that be a more generic solution?

**Comment by draganescu** [comment]
@manooweb Yes, I understand the need and I do think this conversion should be changeable by extenders. My worry is about this solution because it is unique in the block library and because it may impact future updates to the block.

**Comment by adamziel** [comment]
> All my concerns are summed up by just being worried that by adding a filter on client side conversion we make a contract about the structure of the navigation block to the world.

Thanks for bringing it up @draganescu. I agree, I even wanted to explore backend-only conversion to blocks for 6.0. The usual way of navigating the uncertainty is to call things `__unstable`, so I renamed the filter to reflect that.

File: packages/block-library/src/navigation/menu-items-to-blocks.js
Comment (adamziel) [reply packages/block-library/src/navigation/menu-items-to-blocks.js]: Then every implementation would need to access or re-implement `createDataTree` and `mapMenuItemsToBlocks`, plus it would be hard to stack multiple filters.

**Review by draganescu — APPROVED** [review]
With an unstable label this can move forward. I think we'll be better off in the future to remove client side filtering and instead allow extenders to append detail to the output of the menu items endpoint which instructs the conversion, or do the conversion itself serverside.

**Comment by getdave** [comment]
Let's create a follow up Issue so we don't forgot to explore this.

@manooweb this change is marked unstable so the implementation may change in the future (after WP 5.9). I recommend keeping an eye on developments in https://github.com/WordPress/gutenberg/issues/35521.

**Comment by manooweb** [comment]
@draganescu Yes! I understand too and prepare our release with the new unstable filter name.
@getdave I don't see any subject about this in the issue you linked. Did I miss something?
However I already subscribed to it to further follow the developments and I'm going to see if I can set up a test in our side to detect the future changes as soon as possible.

Thanks all for the reactivity and by backported this in the beta/RC releases 👍

**Comment by spacedmonkey** [comment]
Is this ticket not being backported?

**Comment by adamziel** [comment]
I think it is here https://github.com/WordPress/wordpress-develop/pull/2048

**Pull request merged**
